src/Microsoft.ML.CpuMath/AvxIntrinsics.cs

```diff
@@ -173,118 +193,29 @@ public static unsafe void MatMul(ReadOnlySpan<float> mat, ReadOnlySpan<float> sr
-                            // Handle all the 256-bit blocks that we can now that we have offset to an aligned address
-                            remainder = length % 8;
-
-                            while (pSrcCurrent + 8 <= pSrcEnd)
-                            {
-                                // If we aren't using the VEX-encoding, the JIT will only fold away aligned loads
-                                // (due to semantics of the legacy encoding).
-                                // We don't need an assert, since the instruction will throw for unaligned inputs.
-                                Vector256<float> vector = Avx.LoadVector256(pSrcCurrent);
-
-                                float* pMatTemp = pMatCurrent;
-                                res0 = MultiplyAdd(pMatTemp, vector, res0);
-                                res1 = MultiplyAdd(pMatTemp += ccol, vector, res1);
-                                res2 = MultiplyAdd(pMatTemp += ccol, vector, res2);
-                                res3 = MultiplyAdd(pMatTemp += ccol, vector, res3);
-
-                                pSrcCurrent += 8;
-                                pMatCurrent += 8;
-                            }
-                        }
-                        else
-                        {
-                            // Handle the "worst-case" scenario, which is when we have 8-16 elements and the input is not
-                            // 256-bit aligned. This means we can't do any aligned loads and will just end up doing two
-                            // unaligned loads where we mask the input each time.
-                            remainder = length;
-                        }
-
-                        if (remainder != 0)
-                        {
-                            // Handle any trailing elements that don't fit into a 256-bit block by moving back so that the next
-                            // unaligned load will read to the end of the array and then mask out any elements already processed
-
-                            pMatCurrent -= (8 - remainder);
-                            pSrcCurrent -= (8 - remainder);
-
-                            Vector256<float> mask = Avx.LoadVector256(((float*)(pTrailingAlignmentMask)) + (remainder * 8));
+                        float* pMatTemp = pMatCurrent;
 
-                            float* pMatTemp = pMatCurrent;
-                            Vector256<float> x01 = Avx.And(mask, Avx.LoadVector256(pMatTemp));
-                            Vector256<float> x11 = Avx.And(mask, Avx.LoadVector256(pMatTemp += ccol));
-                            Vector256<float> x21 = Avx.And(mask, Avx.LoadVector256(pMatTemp += ccol));
-                            Vector256<float> x31 = Avx.And(mask, Avx.LoadVector256(pMatTemp += ccol));
-                            Vector256<float> vector = Avx.And(mask, Avx.LoadVector256(pSrcCurrent));
+                        Vector256<float> x01 = Avx.LoadAlignedVector256(pMatTemp);
```

Comment:
nit: Codegen will be better on modern computers if we use `LoadVector256` (which does an unaligned load and allows it to be folded into the consuming instruction).

Reply:
NOTE: Changing this to `LoadVector256` will also no longer throw if the input is unaligned, so adding a `Debug.Assert` or a helper might be beneficial. (I have a proposal on CoreFX to add a built-in helper that does this).